src/anomalib/data/dataclasses/generic.py

```diff
@@ -656,5 +658,20 @@ def batch_size(self) -> int:
     def collate(cls: type["BatchIterateMixin"], items: list[ItemT]) -> "BatchIterateMixin":
         """Convert a list of DatasetItem objects to a Batch object."""
         keys = [key for key, value in asdict(items[0]).items() if value is not None]
+
+        # Check if all images have the same shape. If not, resize before collating
+        im_shapes = torch.vstack([tensor(item.image.shape) for item in items if item.image is not None])[..., 1:]
+        if torch.unique(im_shapes, dim=0).size(0) != 1:  # check if batch has heterogeneous shapes
+            target_shape = im_shapes[
+                torch.unravel_index(im_shapes.argmax(), im_shapes.shape)[0],
+                :,
+            ]  # shape of image with largest H or W
+            for item in items:
+                for key in keys:
+                    value = getattr(item, key)
+                    if isinstance(value, Image | Mask):
+                        setattr(item, key, resize(value, target_shape))
```

Comment:
the image transforms are supposed to be applied before the batch collate function right? So this resize operation will only apply if no resizing is defined in the image transforms?

Reply:
With the changes in this PR, the transforms are actually applied by the new `PreProcessor` class using its `on_before_<stage>_batch` hooks, so after collating. This has the following advantages:
- The transforms are contained within the `PreProcessor`. Previously, we needed complicated logic to ensure that the transforms in the datamodule/dataloaders were equal to the transforms stored in (or defined by) the model. Now, the Preprocessor is the only class concerned with the transforms, leading to a much better separation of responsibilities.
- The `PreProcessor` is a submodule of the Lightning model, so we have direct access to the transforms from within the model. This helps with model saving/loading (transforms are automatically stored in the model's state dict), and exporting (transforms are included in the forward pass so that they end up in the model graph).

The only reason why we need resizing in the collate method is to allow collating image batches with heterogeneous image sizes. We can only stack a sequence of image tensors into a single tensor if all images have the same shape. This is why we check the image shapes in the batch, and if they are not equal, we resize all of them to the shape of the image with the biggest width or height. This is only to ensure that we can succesfully collate the images into a single batch, and is completely independent of any further resizing defined in the transforms. (In general, it is recommended to apply resizing in the image transforms, and most of our models include this in their default set transforms).

Reply:
Yes I see! Resizing can be a critical function for image processing models when using different interpolations modes.
What will happen if a user uses a different interpolation mode for their custom transform function:
- the batches with different shaped images are resized in the collate function via the (default) bilinear transform, and later resized using the users custom interpolation mode
- the batches with equal shaped images are only resized using the users custom interpolation mode

There's a good chance that this won't affect anything in anomalib but if accuracy of models drop this is also some place to look after..

Reply:
We hadn't considered that, thanks for pointing it out!

We could potentially add an optional resize argument to the `AnomalibDatamodule` to give the user control over the resizing before collating. But this would require some further investigation and internal discussion.

Reply:
Ok! Yes maybe its not even a problem for the model accuracy. Still it's a bit computational expensive to have one extra resizing operation..